src/nbl/ext/FFT/FFT.cpp

```diff
@@ -0,0 +1,368 @@
+#endif
+
+#endif
+
+// Output Descriptor
+
+struct nbl_glsl_ext_FFT_output_t
+{
+	vec2 complex_value;
+};
+
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_ 1
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+layout(set=_NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_) restrict buffer OutputBuffer
+{
+	nbl_glsl_ext_FFT_output_t outData[];
+};
+#endif
+
+// Get/Set Data Function
+
+vec2 nbl_glsl_ext_FFT_getData(in uvec3 coordinate, in uint channel)
+{
+	vec2 retValue = vec2(0, 0);
+#if USE_SSBO_FOR_INPUT > 0
+	uvec3 dimension = pc.dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	retValue = inData[index].complex_value;
+#else
+	vec4 texelValue= texelFetch(inputImage, ivec2(coordinate.xy), 0);
+	retValue = vec2(texelValue[channel], 0.0f);
+#endif
+	return retValue;
+}
+
+void nbl_glsl_ext_FFT_setData(in uvec3 coordinate, in uint channel, in vec2 complex_value)
+{
+	uvec3 dimension = pc.padded_dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	outData[index].complex_value = complex_value;
+}
+
+void main()
+{
+	nbl_glsl_ext_FFT();
+}
+
+)===";
+
+	const size_t extraSize = 32 + 32 + 32 + 32;
+
+	const uint32_t maxItemsPerThread = core::ceil(float(maxPaddedDimensionSize) / DEFAULT_WORK_GROUP_X_DIM);
+	const uint32_t useSSBO = (DataType::SSBO == inputType) ? 1 : 0;
+	auto shader = core::make_smart_refctd_ptr<ICPUBuffer>(strlen(sourceFmt)+extraSize+1u);
+	snprintf(
+		reinterpret_cast<char*>(shader->getPointer()),shader->getSize(), sourceFmt,
+		useSSBO,
+		DEFAULT_WORK_GROUP_X_DIM,
+		maxPaddedDimensionSize,
+		maxItemsPerThread
+	);
+
+	auto cpuSpecializedShader = core::make_smart_refctd_ptr<ICPUSpecializedShader>(
+		core::make_smart_refctd_ptr<ICPUShader>(std::move(shader),ICPUShader::buffer_contains_glsl),
+		ISpecializedShader::SInfo{nullptr, nullptr, "main", asset::ISpecializedShader::ESS_COMPUTE}
+	);
+	
+	auto gpuShader = driver->createGPUShader(nbl::core::smart_refctd_ptr<const ICPUShader>(cpuSpecializedShader->getUnspecialized()));
+	
+	auto gpuSpecializedShader = driver->createGPUSpecializedShader(gpuShader.get(), cpuSpecializedShader->getSpecializationInfo());
+
+	return gpuSpecializedShader;
+}
+
+void FFT::defaultBarrier()
+{
+	COpenGLExtensionHandler::pGlMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT);
+}
+
+// Kernel Normalization
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> FFT::createKernelNormalizationShader(video::IVideoDriver* driver)
+{
+	const char* sourceFmt =
```

Comment:
we avoid having GLSL in C++ source as strings like the plague

please move these to `nbl/builtin/glsl/ext/FFT/something.glsl` befrore I review

Reply:
since its supposed to be a compute shader then it could have the `.comp` extension